assets/js/phoenix.js

```diff
@@ -940,12 +940,46 @@ export class Socket {
   }
 
   teardown(callback, code, reason){
-    if(this.conn){
-      this.conn.onclose = function(){} // noop
-      if(code){ this.conn.close(code, reason || "") } else { this.conn.close() }
-      this.conn = null
+    if (!this.conn) {
+      return callback && callback()
     }
-    callback && callback()
+
+    this.waitForBufferDone(() => {
+      if (this.conn) {
+        if(code){ this.conn.close(code, reason || "") } else { this.conn.close() }
+      }
+
+      this.waitForSocketClosed(() => {
+        if (this.conn) {
+          this.conn.onclose = function(){} // noop
+          this.conn = null
+        }
+
+        callback && callback()
+      })
+    })
+  }
+
+  waitForBufferDone(callback, tries = 1) {
+    if (tries === 5 || !this.conn || (this.conn.bufferedAmount && this.conn.bufferedAmount === 0)) {
+      callback()
+      return
+    }
+
+    setTimeout(() => {
+      this.waitForBufferDone(callback, tries + 1)
+    }, 150 * tries)
```

Comment:
can you name 150 and 5 above, so they'd be less magic numbers

Reply:
Also a great idea but b/c this is just one large file, creating const's is much less appealing.

Reply:
it could be in the same function as an assignment like `const maxTriesCount = 5`. but I guess it's fine as now as well. My only concern is that when changing code and I see timeout with number 150, then I have no idea how was this chosen and if it is safe to change. I usually add some information in the commit message or as a code comment how this was chosen.